sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3894,16 +3899,99 @@ pi_result piextUSMSharedAlloc(void **ResultPtr, pi_context Context,
   ZeDevDesc.ordinal = 0;
   ZE_CALL(zeDriverAllocSharedMem(Context->Device->Platform->ZeDriver,
                                  &ZeDevDesc, &ZeHostDesc, Size, Alignment,
-                                 Device->ZeDevice, ResultPtr));
+                                 Context->Device->ZeDevice, ResultPtr));
 
   return PI_SUCCESS;
 }
 
-pi_result piextUSMFree(pi_context Context, void *Ptr) {
+pi_result USMFreeImpl(pi_context Context, void *Ptr) {
   ZE_CALL(zeDriverFreeMem(Context->Device->Platform->ZeDriver, Ptr));
   return PI_SUCCESS;
 }
 
+// Exception type to pass allocation errors
+class UsmAllocationException {
+  const pi_result Error;
+public:
+  UsmAllocationException(pi_result Err): Error{Err} {}
+  pi_result getError() const { return Error; }
+};
+
+void* USMSharedMemoryAlloc::allocate(size_t Size) {
+  void* Ptr = nullptr;
+
+  // TODO: how to deal with properties? Allocate a specific slab or just bypass the allocator?
+  // They seems to be not used by DPC++ RT right now.
+  auto Res = USMSharedAllocImpl(&Ptr, Context, nullptr, Size, sizeof(void*));
+  if (Res != PI_SUCCESS) {
+    throw UsmAllocationException(Res);
+  }
+
+  return Ptr;
+}
+
+void* USMSharedMemoryAlloc::allocate(size_t Size, size_t Alignment) {
+  void* Ptr = nullptr;
+
+  auto Res = USMSharedAllocImpl(&Ptr, Context, nullptr, Size, Alignment);
+  if (Res != PI_SUCCESS) {
+    throw UsmAllocationException(Res);
+  }
+  return Ptr;
+}
+
+void USMSharedMemoryAlloc::deallocate(void* Ptr) {
+  USMFreeImpl(Context, Ptr);
+}
+
+pi_result piextUSMSharedAlloc(void **ResultPtr, pi_context Context,
+                              pi_device Device,
+                              pi_usm_mem_properties *Properties, size_t Size,
+                              pi_uint32 Alignment) {
+  (void) Device;
+  if (!UseUSMAllocator ||
+     // L0 spec says that allocation fails if Alignment != 2^n, in order to keep
```

Comment:
I don't see why DPC++ should continue to fail on non power-2 alignment if USM allocator can gracefully handle this.

Reply:
The idea was to align the behavior with L0 runtime as it's unclear what we should do here: L0 runtime itself gracefully handle such alignment(basically ignores it and uses default large value), but at the same time: L0 spec prohibits it and running such code via L0 validation level returns an error.